subtype/src/test/java/com/fasterxml/jackson/module/subtype/WithJsonSubTypesTest.java

```diff
@@ -0,0 +1,273 @@
+package com.fasterxml.jackson.module.subtype;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.auto.service.AutoService;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import static org.junit.Assert.*;
+
+/**
+ * test work with {@link JsonSubTypes}
+ */
+@RunWith(value = Parameterized.class)
+public class WithJsonSubTypesTest<T extends WithJsonSubTypesTest.Parent> {
+
+    private final ObjectMapper mapper = new ObjectMapper().findAndRegisterModules();
+
+    public static class Argument<T> {
+        private final Class<T> clazz;
+        private final T expected;
+
+        public Argument(Class<T> clazz, T expected) {
+            this.clazz = clazz;
+            this.expected = expected;
+        }
+    }
+
+    @Parameter
+    public Argument<T> argument;
+
+    @Parameters
+    public static Collection<Argument<?>> data() {
+        return Arrays.asList(
+                new Argument<>(FirstChild.class, new FirstChild("hello")),
+                new Argument<>(SecondChild.class, new SecondChild("world")),
+                new Argument<>(FirstAppendChild.class, new FirstAppendChild(42)),
+                new Argument<>(SecondAppendChild.class, new SecondAppendChild("42", Arrays.asList("hello", "foo", "bar"))),
+                new Argument<>(ThirdAppendChild.class, new ThirdAppendChild("42", Arrays.asList("hello", "foo", "bar"), 3.1415926))
+        );
+    }
+
+    @Test
+    public void test() throws Exception {
+        final Parent parent = argument.expected;
+        String json = mapper.writeValueAsString(parent);
+        Parent unmarshal = mapper.readValue(json, Parent.class);
+        T actual = assertInstanceOf(argument.clazz, unmarshal);
+        assertEquals(argument.expected, actual);
+    }
+
+    public static <T> T assertInstanceOf(Class<T> expectedType, Object actualValue) {
+        assertTrue(expectedType.isInstance(actualValue));
+        return expectedType.cast(actualValue);
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+    @JsonSubTypes(value = {
+            @JsonSubTypes.Type(value = FirstChild.class, name = "first-child"),
+            @JsonSubTypes.Type(value = SecondChild.class, name = "second-child"),
+    })
+    public interface Parent {
+    }
+
+    public static class FirstChild implements Parent {
+        private String foo;
+
+        @SuppressWarnings("unused") // SPI require it
+        public FirstChild() {
+        }
+
+        public FirstChild(String foo) {
+            this.foo = foo;
+        }
+
+        @SuppressWarnings("unused") // jackson require it
```

Comment:
We could instead just use `public String foo` to avoid having to add getter and setter